core/modules/drr.cc

```diff
@@ -0,0 +1,392 @@
+      if (err < 0)
+        return err;
+
+      if (llring_empty(f->queue) && !f->next_packet) {
+        // if the flow expired, remove it
+        if (now - f->timer > kTTL) {
+          RemoveFlow(f);
+        } else {
+          err = llring_sp_enqueue(flow_ring_, f);
+          if (err < 0)
+            return err;
+        }
+        continue;
+      }
+
+      f->deficit += quantum_;
+    } else {
+      f = current_flow_;
+      current_flow_ = nullptr;
+    }
+
+    int bytes = GetNextPackets(batch, f);
+    if (bytes < 0) {
+      return bytes;
+    }
+    total_bytes += bytes;
+
+    if (llring_empty(f->queue) && !f->next_packet) {
+      f->deficit = 0;
+    }
+
+    // if the flow doesn't have any more packets to give, reenqueue it
+    if (!f->next_packet || f->next_packet->total_len() > f->deficit) {
+      err = llring_sp_enqueue(flow_ring_, f);
+      if (err < 0)
+        return err;
+    } else  {
+      // knowing that the while statement will exit, keep the flow that still
+      // has packets at the front
+      current_flow_ = f;
+    }
+  }
+  return total_bytes;
+}
+
+int DRR::GetNextPackets(bess::PacketBatch* batch, Flow* f) {
+  int total_bytes = 0;
+  bess::Packet* pkt;
+
+  while (!batch->full() && !llring_empty(f->queue)) {
+    // makes sure there isn't already a packet at the front
+    if (!f->next_packet) {
+      int err = llring_sc_dequeue(f->queue, reinterpret_cast<void**>(&pkt));
+      if (err < 0)
+        return err;
+    } else {
+      pkt = f->next_packet;
+      f->next_packet = nullptr;
+    }
+
+    if (pkt->total_len() > f->deficit) {
+      f->next_packet = pkt;
+      break;
+    }
+
+    f->deficit -= pkt->total_len();
+    total_bytes += pkt->total_len();
+    batch->add(pkt);
+  }
+
+  return total_bytes;
+}
+
+DRR::Flow_Id DRR::GetId(bess::Packet* pkt) {
+  struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+  struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+  int ip_bytes = (ip->version_ihl & 0xf) << 2;
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(
+      reinterpret_cast<uint8_t *>(ip) + ip_bytes);  // Assumes a l-4 header
+  // TODO(joshua): handle packet fragmentation
+  Flow_Id id = {ip->src_addr, ip->dst_addr, udp->src_port,
+      udp->dst_port, ip->next_proto_id};
+  return id;
+}
+
+int DRR::AddNewFlow(bess::Packet *pkt, Flow_Id id) {
+  int err;
+
+  // creates flow
+  Flow* f = static_cast<Flow *>(mem_alloc_ex(sizeof(Flow), alignof(Flow), 0));
+  Flow init = {0, 0, id, 0, nullptr};
+  memcpy(f, &init, sizeof(init));
+
+  // TODO(joshua) do proper error checking
+  err = AddQueue(&f->queue, static_cast<int>(kFLOW_QUEUE_SIZE));
+
+  if (err != 0)
+    return err;
+
+  flows_.Insert(id, f);
+
+  err = Enqueue(f, pkt);
+  if (err != 0)
+    return err;
+
+  // puts flow in round robin
+  err = llring_sp_enqueue(flow_ring_, f);
+  if (err != 0)
+    return err;
+
+  return 0;
+}
+
+void DRR::RemoveFlow(Flow* f) {
```

Comment:
Much of this could be placed in Flow's destructor, if you were to define one.  The only parts that would need to be outside would be parts that change flows_